lib/shopify_app/session/jwt.rb

```diff
@@ -0,0 +1,46 @@
+module ShopifyApp
+  class JWT
+    def initialize(token)
+      @token = token
+      set_payload
+    end
+
+    def shopify_domain
+      payload && dest_host
+    end
+
+    def shopify_user_id
+      payload && payload['sub']
+    end
+
+    private
+
+    def payload
+      return unless @payload
+      return unless dest_host
+      return unless dest_host == iss_host
+      return unless @payload['aud'] == ShopifyApp.configuration.api_key
+
+      @payload
+    end
+
+    def set_payload
+      parse_token_data(ShopifyApp.configuration.secret)
+      parse_token_data(ShopifyApp.configuration.old_secret) if !@payload && ShopifyApp.configuration.old_secret
+    end
+
+    def parse_token_data(secret)
+      @payload, _ = ::JWT.decode(@token, secret, true, { algorithm: 'HS256' })
```

Comment:
I would move the `@payload =` setter into `set_payload` instead of in this method. This method (by its name) should be focused on parsing and the `set_payload` method can actually do the work of setting the instance variable.